src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer.Xaml/Handler/Completion/CompletionHandler.cs

```diff
@@ -0,0 +1,198 @@
+            // If we have already cached the commit characters, return the cached ones
+            return cachedCharacters.ToArray();
+        }
+
+        var xamlCommitCharacters = completionItem.XamlCommitCharacters;
+
+        var commitCharacters = xamlCommitCharacters.Characters.Select(c => new VSInternalCommitCharacter { Character = c.ToString(), Insert = !xamlCommitCharacters.NonInsertCharacters.Contains(c) }).ToImmutableArray();
+        commitCharactersCache.Add(completionItem.Kind, commitCharacters);
+        return commitCharacters.ToArray();
+    }
+
+    private static CompletionItemKind GetItemKind(XamlCompletionKind kind)
+    {
+        switch (kind)
+        {
+            case XamlCompletionKind.Element:
+            case XamlCompletionKind.ElementName:
+                return CompletionItemKind.Element;
+            case XamlCompletionKind.EndTag:
+                return CompletionItemKind.CloseElement;
+            case XamlCompletionKind.Attribute:
+            case XamlCompletionKind.AttachedPropertyValue:
+            case XamlCompletionKind.ConditionalArgument:
+            case XamlCompletionKind.DataBoundProperty:
+            case XamlCompletionKind.MarkupExtensionParameter:
+            case XamlCompletionKind.PropertyElement:
+                return CompletionItemKind.Property;
+            case XamlCompletionKind.ConditionValue:
+            case XamlCompletionKind.MarkupExtensionValue:
+            case XamlCompletionKind.PropertyValue:
+            case XamlCompletionKind.Value:
+                return CompletionItemKind.Value;
+            case XamlCompletionKind.Event:
+            case XamlCompletionKind.EventHandlerDescription:
+                return CompletionItemKind.Event;
+            case XamlCompletionKind.NamespaceValue:
+            case XamlCompletionKind.Prefix:
+                return CompletionItemKind.Namespace;
+            case XamlCompletionKind.AttachedPropertyTypePrefix:
+            case XamlCompletionKind.MarkupExtensionClass:
+            case XamlCompletionKind.Type:
+            case XamlCompletionKind.TypePrefix:
+                return CompletionItemKind.Class;
+            case XamlCompletionKind.LocalResource:
+                return CompletionItemKind.LocalResource;
+            case XamlCompletionKind.SystemResource:
+                return CompletionItemKind.SystemResource;
+            case XamlCompletionKind.CData:
+            case XamlCompletionKind.Comment:
+            case XamlCompletionKind.ProcessingInstruction:
+            case XamlCompletionKind.RegionStart:
+            case XamlCompletionKind.RegionEnd:
+                return CompletionItemKind.Keyword;
+            case XamlCompletionKind.Snippet:
+                return CompletionItemKind.Snippet;
+            default:
+                Debug.Fail($"Unhandled {nameof(XamlCompletionKind)}: {Enum.GetName(typeof(XamlCompletionKind), kind)}");
+                return CompletionItemKind.Text;
+        }
+    }
+}
```

Comment:
i don't love all this code now living in roslyn.  can these not live elsewhere, just enabled through an EA layer?